# WARNING: this file is auto-generated by 'build_sync_library.py'
# from the original file 'test_mongodb.py'
# DO NOT CHANGE! Change the original file instead.
import contextlib
from collections.abc import Generator
from typing import Any

import pytest
from inline_snapshot import snapshot
from pymongo import MongoClient
from typing_extensions import override

from key_value.sync.code_gen.stores.base import BaseStore
from key_value.sync.code_gen.stores.mongodb import MongoDBStore
from tests.code_gen.conftest import docker_container
from tests.code_gen.stores.conftest import BaseStoreTests, ContextManagerStoreTestMixin, should_skip_docker_tests, wait_for_store

# MongoDB test configuration
MONGODB_HOST = "localhost"
MONGODB_HOST_PORT = 27017
MONGODB_TEST_DB = "kv-store-adapter-tests"

WAIT_FOR_MONGODB_TIMEOUT = 30


def ping_mongodb() -> bool:
    try:
        client: MongoClient[Any] = MongoClient[Any](host=MONGODB_HOST, port=MONGODB_HOST_PORT)
        _ = client.list_database_names()
    except Exception:
        return False

    return True


class MongoDBFailedToStartError(Exception):
    pass


@pytest.mark.skipif(should_skip_docker_tests(), reason="Docker is not available")
class TestMongoDBStore(ContextManagerStoreTestMixin, BaseStoreTests):
    @pytest.fixture(autouse=True, scope="session")
    def setup_mongodb(self) -> Generator[None, None, None]:
        with docker_container("mongodb-test", "mongo:7", {"27017": 27017}):
            if not wait_for_store(wait_fn=ping_mongodb):
                msg = "MongoDB failed to start"
                raise MongoDBFailedToStartError(msg)

            yield

    @override
    @pytest.fixture
    def store(self, setup_mongodb: None) -> MongoDBStore:
        store = MongoDBStore(url=f"mongodb://{MONGODB_HOST}:{MONGODB_HOST_PORT}", db_name=MONGODB_TEST_DB)
        # Ensure a clean db by dropping our default test collection if it exists
        with contextlib.suppress(Exception):
            _ = store._client.drop_database(name_or_database=MONGODB_TEST_DB)  # pyright: ignore[reportPrivateUsage]

        return store

    @pytest.fixture
    def mongodb_store(self, store: MongoDBStore) -> MongoDBStore:
        return store

    @pytest.mark.skip(reason="Distributed Caches are unbounded")
    @override
    def test_not_unbounded(self, store: BaseStore): ...

    def test_mongodb_collection_name_sanitization(self, mongodb_store: MongoDBStore):
        """Tests that a special characters in the collection name will not raise an error."""
        mongodb_store.put(collection="test_collection!@#$%^&*()", key="test_key", value={"test": "test"})
        assert mongodb_store.get(collection="test_collection!@#$%^&*()", key="test_key") == {"test": "test"}

        collections = mongodb_store.collections()
        assert collections == snapshot(["test_collection_-daf4a2ec"])
